### DISCUSSION TITLE
How would you validate individual array elements, for example, using the Distinct rule?

### DISCUSSION THREAD

**Discussion by ragulka** [spatie/laravel-data#222] [Q&A] [answered]
Given an array, I want to ensure it only contains distinct values. Using regular Laravel validation rules, I can define the validation as the following. Even though slightly confusing (I would expect I can assign the `distinct` rule directly to the `tags` attribute, instead of having to define it for each element in the tags array) - it works:

```php
Validator::make([
    'tags' => ['foo', 'foo', 'bar']
], [
    'tags' => 'array',
    'tags.*' => 'distinct',
])->errors();

// this results in something like:
[
    "tags.0" => [
      "The tags.0 field has a duplicate value.",
    ],
    "tags.1" => [
      "The tags.1 field has a duplicate value.",
    ],
]
```

How can I replicate the same using a Data Object and validation attributes? FWIW, the following does not work:

```php
class MyData extends Data
{
    public function __construct(
        #[Distinct]
        public array $tags,
    ) {
    }
}
```

**Comment by rubenvanassche** [comment, marked as answer]
At the moment the best solution is to manually add the distinct rule using a static `rules` method. Our current architecture doesn't supporting adding rules like `'tags.*' => 'distinct'`. I'm taking a look into how we can change this but it will be quite a bit of work.
